core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -720,6 +691,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     } finally {
       endProcessing(rootPath)
       pendingReplayTasksCount.decrementAndGet()
+
+      // triggering another task for compaction task
+      try {
+        processing(rootPath)
+        val task: Runnable = () => compact(reader)
+        replayExecutor.submit(task)
+      } catch {
+        // let the iteration over the updated entries break, since an exception on
+        // replayExecutor.submit (..) indicates the ExecutorService is unable
+        // to take any more submissions at this time
+        case e: Exception =>
+          logError(s"Exception while submitting task for compaction", e)
+          endProcessing(rootPath)
```

Comment:
I think checkLog seems to have a bug - if replayExecutor.submit fails and throws an exception, the log is not removed from `processing` which would block processing for the log in both `checkLogs` and `cleanLogs`. 

I'm OK to fix it here as well, or fix it in another PR to avoid mixing up.

Reply:
>  if replayExecutor.submit fails 

That should be extremely rare, but seems simple to handle.

Reply:
Ah yes that's being handled as I extracted method and addressed there.